### DISCUSSION TITLE
question about async init function

### DISCUSSION THREAD

**Discussion by Chandler-Zhu** [alpinejs/alpine#2226] [1. Help] [answered]
I tried to use async/await syntax in the init function, but when I run the code it gives me script error
https://jsfiddle.net/uot2943q/1/

**Comment by SimoTod** [comment, marked as answer]
There are several issues in your example (You probably want to read some tutorials about async/await and variable scoping):
- you missed the closing } for x-data
- you didn't define `result` in x-data but you are trying to use it in x-text
- setTimeout doesn't return a promise therefore awaiting setTimeout doesn't make sense
- For assignments, the syntax is `variable = await promise` not `await variable = promise`
- you are defining a and b as parameters in the setTimeout so it won't use the variable defined outside 

Basically, your example should be changed to something like that: https://codepen.io/SimoTod/pen/qBXbabg

**Reply by Chandler-Zhu** [reply]
I did some reading but there is still one thing that doesn't make sense to me, at the end of `async init` function, why is it `this.result = await promise;` instead of `this.result = await promise.result;`  ?

**Reply by SimoTod** [reply]
It's just the name I gave to the variable, you can change it to something else like
```javascript
   ...
   let foo = new Promise((resolve) => {
      setTimeout(() => {
        resolve(a + b)
      }, 3000);
    })
    this.result = await foo;
```
Basically, we create a variable called foo and we assign a Promise object to it. Our promise object internally uses a setTimeout to simulate a delay and then resolve to a+b. This means that foo is a 'pending promise' (aka not good for any use yet) for 3 seconds then it becomes 12.
the last line assigns foo to result but we use the keyword away to tell the javascript engine to "wait" until the promise is resolved (aka becomes 12) before doing the assignment.
At that point, because you are changing a property of x-data, Alpine will update the page for you.
I hope it makes sense

**Reply by Chandler-Zhu** [reply]
An async function always returns a promise, in this case what does `asycn init` return ? Is it the promise named `foo` or the `result` wrapped in a promise ?

**Reply by SimoTod** [reply]
it doesn't return anything (or if you like it implicitly returns a promise already resolved that doesn't do anything). 

We use async there because you can't use `await` otherwise. 
The returned value of init() is not used anywhere so it doesn't really matter what it returns.

`foo` is a different promise that you use inside init().

It's not an entry-level argument but there should be a lot of tutorials and examples available on the web.
